### ISSUE TITLE
ConnectionManager.ConnectToPeer creates a PeerId with outdated BitField

### ISSUE THREAD

**Issue by ilyalatt** [mono/monotorrent#196]
Hello. I initialize `TorrentManager` with `MagnetLink`. I wait until `TorrentManager.Torrent != null`. Connected peers have `BitField` with single element. After some time bitfields are fixed. `IgnoringPicker` throws on `BitField.Check` meahwhile. This bug can be fixed in `MetadataMode.SwitchToRegular` but I am not sure it is a good way.

**Comment by alanmcgovern** [comment]
When in `MetadataMode` the bitfield for the TorrentManager is used to represent the .torrent metadata. When the metadata (the .torrent file) has been fully downloaded the Bitfield is replaced with the one corresponding to the actual data and all peers are reconnected.

What exception is being thrown by the piece picker? Do you have the stacktrace for the error? What makes the BitField 'outdated'?

**Comment by ilyalatt** [comment]
> What makes the BitField 'outdated'?

I meant that BitField is not updated. I called initial value 'F' of `PeerId.BitField` as 'oudated'.

>  Do you have the stacktrace for the error?

```
Unexpected main loop exception: System.ArgumentException: BitFields are of different lengths (Parameter 'value')
   at MonoTorrent.BitField.Check(BitField value)
   at MonoTorrent.BitField.From(BitField value)
   at MonoTorrent.Client.PiecePicking.IgnoringPicker.PickPiece(IPieceRequester peer, BitField available, IReadOnlyList`1 otherPeers, Int32 count, Int32 startIndex, Int32 endIndex)
   at MonoTorrent.Client.PieceManager.AddPieceRequests(PeerId id)
   at MonoTorrent.Client.ChokeUnchokeManager.UnchokeReview()
   at MonoTorrent.Client.Modes.Mode.DownloadLogic(Int32 counter)
   at MonoTorrent.Client.Modes.Mode.Tick(Int32 counter)
   at MonoTorrent.Client.Modes.DownloadMode.Tick(Int32 counter)
   at MonoTorrent.Client.ClientEngine.LogicTick()
   at MonoTorrent.Client.ClientEngine.<.ctor>b__66_0()
   at MonoTorrent.Client.MainLoop.<>c__DisplayClass10_0.<QueueTimeout>b__0(Object state)
   at MonoTorrent.Client.MainLoop.Loop()
```

> When the metadata (the .torrent file) has been fully downloaded the Bitfield is replaced with the one corresponding to the actual data and all peers are reconnected.

Here are my initial thoughts. `Mode.DownloadLogic` copies `Manager.BitField`. It is ok if a metadata is present. It is not ok if a metadata is not present. `MetadataMode.SwitchToRegular` calls `CleanupSocket` for `PeerId` but does not update `PeerId.BitField`. Looks like `Mode.HandleBitfieldMessage` should set correct `PeerId.BitField` for other PeerIds. However it is not happening immediately so I get the exception.

But I have discovered that only one PeerId among 8 PeerIds has 'outdated' BitField. So I do not have ideas what is going on.

**Comment by alanmcgovern** [comment]
I wonder if this is a mostly harmless race condition.

When swapping between MetadataMode and regular downloading/seeding mode all pre-existing connections are closed. I can't remember *why* they're closed instead of being updated, but let's assume there's a good reason ;p I can review that later. As such, the expectation is that 'metadata' peers are connected when in metadata mode, and 'regular' peers are connected in regular mode.

There are two types of connected peer, there's the peer which has fully completed the handshaking and the peer which has not yet done so. I can't remember precisely why these two cases are distinguished (hopefully version control history might explain).

If a new connection is in the middle of handshaking and we swap to regular mode, it will not be closed. That will cause a peer with the wrong bitfield to be added to the 'ConnectedPeers' list.

Could you check if this branch prevents the issue? https://github.com/mono/monotorrent/tree/fix-196 . Double points if you can confirm that the new `TorrentException` is actually being thrown.

If you're not compiling from source I can produce a nuget to help you test.

**Comment by ilyalatt** [comment]
I have checked your fix a few times. It works. Thank you.

**Comment by alanmcgovern** [comment]
Amazing, thanks for checking that so quickly.

I'll review the usage of the 'HandshakingPeers' list as I don't believe it makes sense to separate these peers from the rest of the 'ConnectedPeers'. The change went in as part of a large refactor ( https://github.com/mono/monotorrent/commit/b4ee76b33a32698a79bcde8204f4ae328e9da04b ) and there's no specific reason given for it. I can't think of any reason why this would be needed.

**Comment by alanmcgovern** [comment]
I'll leave this open until I investigate/implement one of the other options :) It'll help me not forget!

Also, i just released 1.0.10 in NuGet which has this patch, so you can update your nuget package or just pull the latest code from master if you want this change!

**Comment by ilyalatt** [comment]
Yeah. I updated the package as soon as you released it :) Here is [my project](https://github.com/ilyalatt/BitTorrentStorage). I think it can be referenced in README as an example of fetching custom pieces.

**Comment by ilyalatt** [comment]
Let me elaborate on this. I mean it is hard to find any examples for the library. There is `ClientSample.cs` but it does not answer the question how to fetch custom pieces. Even `SlidingWindowPicker` is not there. I have found about it from [11 years old post](https://bittorrent.ibiblio.narkive.com/J397Cxxp/piece-download-strategy#post5). I spent a few hours to find out how to interact with the library for the purpose of custom piece fetching. BitTorrentStorage is a hacky solution but I guess it can be useful for the people trying to build something similar.
